cli/command/inspect/inspector.go

```diff
@@ -38,7 +38,7 @@ func NewTemplateInspector(outputStream io.Writer, tmpl *template.Template) Inspe
 // NewTemplateInspectorFromString creates a new TemplateInspector from a string
 // which is compiled into a template.
 func NewTemplateInspectorFromString(out io.Writer, tmplStr string) (Inspector, error) {
-	if tmplStr == "" {
+	if tmplStr == "" || tmplStr == "json" {
 		return NewIndentedInspector(out), nil
```

Comment:
This is one we need to have a look at, to see what would be best / make the JSON output most useful.

Currently, using `'{{ json . }}'` as format will not use pretty-print. That makes it hard to read, but potentially (likely?) more useful when processing the output (?)

```
docker inspect --format '{{ json . }}' foo
{"Id":"sha256:bf18dbdb03f7d053c623dcc7ad372dc9a687aa653c08898b5083a8940c29f02d","RepoTags":["foo:latest"],"RepoDigests":[],"Parent":"sha256:491198851f0ccdd0882cb9323f3856043d4e4c65b773e8eac3e0f6bc979a2ae7","Comment":"","Created":"2021-03-10T14:07:37.20034742Z","Container":"","ContainerConfig":{"Hostname":"","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"],"Cmd":["/bin/sh","-c","#(nop) COPY multi:539d0f63df854877d3aad06b13fd267ff9a979d372600ca5075a89f8ffbf4e3e in /work/config/ "],"Image":"sha256:491198851f0ccdd0882cb9323f3856043d4e4c65b773e8eac3e0f6bc979a2ae7","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"DockerVersion":"20.10.2","Author":"","Config":{"Hostname":"","Domainname":"","User":"","AttachStdin":false,"AttachStdout":false,"AttachStderr":false,"Tty":false,"OpenStdin":false,"StdinOnce":false,"Env":["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"],"Cmd":["sh"],"Image":"sha256:491198851f0ccdd0882cb9323f3856043d4e4c65b773e8eac3e0f6bc979a2ae7","Volumes":null,"WorkingDir":"","Entrypoint":null,"OnBuild":null,"Labels":null},"Architecture":"amd64","Os":"linux","Size":1231733,"VirtualSize":1231733,"GraphDriver":{"Data":{"LowerDir":"/var/lib/docker/overlay2/52d1c55b0fb27d947e9298f54d9ffb9e3bcdde45aa8b0b5c2940fb6a8eaa5c30/diff","MergedDir":"/var/lib/docker/overlay2/bde48edd93ce919389f68d17145c534556cc4444d7238477b6f95a4ec70a6c1b/merged","UpperDir":"/var/lib/docker/overlay2/bde48edd93ce919389f68d17145c534556cc4444d7238477b6f95a4ec70a6c1b/diff","WorkDir":"/var/lib/docker/overlay2/bde48edd93ce919389f68d17145c534556cc4444d7238477b6f95a4ec70a6c1b/work"},"Name":"overlay2"},"RootFS":{"Type":"layers","Layers":["sha256:84009204da3f70b09d2be3914e12844ae9db893aa85ef95df83604f95df05187","sha256:ab70170e9d91c94b190c2a3c8a519c24755055a2a52ba40a9ad6cce47bab820a"]},"Metadata":{"LastTagTime":"2021-03-10T14:07:37.293854938Z"}}
```

The _default_ output (no `--format`) on the other hand _does_ use pretty-print (output truncated);

```
docker inspect foo

[
    {
        "Id": "sha256:bf18dbdb03f7d053c623dcc7ad372dc9a687aa653c08898b5083a8940c29f02d",
        "RepoTags": [
            "foo:latest"
        ],
        "RepoDigests": [],
        "Parent": "sha256:491198851f0ccdd0882cb9323f3856043d4e4c65b773e8eac3e0f6bc979a2ae7",
        "Comment": "",
        "Created": "2021-03-10T14:07:37.20034742Z"
    }
]
```

So, I'm a bit on the fence what's most useful;

- do we need separate `json` and `json-pretty` (e.g.) options?
- if we do, do we also need the same in templates (`--format '{{ jsonpretty .Foo.Bar }}'` ?
- does pretty-print make processing the output more _difficult_ to use in scripts?

Reply:
You're right about the pretty vs non-pretty. I'd be in favor of keeping the pretty json print with `no format`, and change here the way we print when `json` is set with non-pretty.
That way the output would be consistent between inspect and list commands.
We could keep the `jsonpretty` format option idea for a follow up eventually, I guess this PR is big enough 😅

Reply:
My two cents,

Parsing JSON shouldn't be different between a pretty-printed and a minimized output. I would like to have all json pretty-printed by default (and only option).

Reply:
> Parsing JSON shouldn't be different between a pretty-printed and a minimized output. I would like to have all json pretty-printed by default (and only option).

JSON is not for humans, so from that perspective, the default should not necessarily be "nicely formatted". To format it nicely, `jq` is a good option (although I wouldn't be fully against having a built-in way to do `pretyjson`). There's some reasons why it's still useful to default to "non-pretty-print";

I agree that either "pretty" or "non-pretty" should work _if you're using a JSON parser_ (which of course is the preferred way to deal with JSON). However, this may not always be what people use, and in some cases may not be needed. It also depends a bit on what kind of JSON (single result, multiple?) which can be the case for _both_ `docker inspect` (`docker inspect <foo> <bar> <baz>`), and by default on (e.g.) `docker ps`.

Here's the JSON output for a `docker ps`;

```bash
docker ps --format '{{ json .}}'
{"Command":"\"/bin/bash\"","CreatedAt":"2022-03-08 09:29:44 +0100 CET","ID":"059a153f844c","Image":"fedora:35","Labels":"maintainer=Clement Verna \u003ccverna@fedoraproject.org\u003e","LocalVolumes":"0","Mounts":"","Names":"stupefied_kirch","Networks":"bridge","Ports":"","RunningFor":"2 hours ago","Size":"553MB (virtual 706MB)","State":"running","Status":"Up 2 hours"}
{"Command":"\"/docker-entrypoint.…\"","CreatedAt":"2022-03-06 18:27:30 +0100 CET","ID":"b778cbd0fc54","Image":"nginx:alpine","Labels":"maintainer=NGINX Docker Maintainers \u003cdocker-maint@nginx.com\u003e","LocalVolumes":"1","Mounts":"myvolume","Names":"foo","Networks":"bridge","Ports":"80/tcp","RunningFor":"41 hours ago","Size":"1.09kB (virtual 22.9MB)","State":"running","Status":"Up 41 hours"}
{"Command":"\"/bin/sh\"","CreatedAt":"2022-03-05 12:24:21 +0100 CET","ID":"5b88bd2598fc","Image":"quay.io/goswagger/swagger:0.7.4","Labels":"","LocalVolumes":"0","Mounts":"","Names":"keen_dubinsky","Networks":"bridge","Ports":"","RunningFor":"2 days ago","Size":"83B (virtual 304MB)","State":"running","Status":"Up 2 days"}
{"Command":"\"/usr/bin/gotty --ti…\"","CreatedAt":"2022-01-06 10:38:22 +0100 CET","ID":"564795c2a515","Image":"thajeztah/ddshell","Labels":"com.docker.desktop.extension.api.version=0.1.0,com.docker.compose.config-hash=4909202d146df86995d6d1e254f94e8f9fd4c084a46ae58c7396d16f81b4bcff,com.docker.compose.container-number=1,com.docker.compose.project=thajeztah_ddshell-desktop-extension,com.docker.compose.project.working_dir=/Users/sebastiaan/Library/Containers/com.docker.docker/Data/plugins/thajeztah_ddshell/vm,com.docker.compose.service=ddshell,com.docker.compose.image=sha256:8fc0de0137378ca03f30d718801d9520c6f0bd323247c5cb7ba26b8b53c24037,com.docker.compose.project.config_files=/Users/sebastiaan/Library/Containers/com.docker.docker/Data/plugins/thajeztah_ddshell/vm/docker-compose.yml,com.docker.compose.version=2.2.2,com.docker.compose.oneoff=False,org.opencontainers.image.description=Docker Desktop Shell Extension,org.opencontainers.image.title=DDShell,com.docker.compose.depends_on=,com.docker.desktop.plugin=true,com.docker.desktop.plugin.name=DDShell,org.opencontainers.image.vendor=Docker Inc.","LocalVolumes":"0","Mounts":"","Names":"thajeztah_ddshell-desktop-extension-service","Networks":"thajeztah_ddshell-desktop-extension_default","Ports":"0.0.0.0:8129-\u003e8080/tcp","RunningFor":"2 months ago","Size":"0B (virtual 15.8MB)","State":"running","Status":"Up 4 days"}
{"Command":"\"/docker-entrypoint.…\"","CreatedAt":"2022-01-05 18:49:16 +0100 CET","ID":"0f1509f6a292","Image":"crazymax/ddext-docs:latest","Labels":"com.docker.compose.config-hash=27fb4a55efe1da12162dddb2ffc7a7799d3efe72a44b1e477bffc68ce4172aa8,com.docker.compose.depends_on=,com.docker.compose.service=devenv-volumes,com.docker.desktop.extension.api.version=0.1.0,maintainer=NGINX Docker Maintainers \u003cdocker-maint@nginx.com\u003e,org.opencontainers.image.description=Docs Extension.,com.docker.compose.container-number=1,com.docker.compose.project=crazymax_ddext-docs-desktop-extension,com.docker.compose.project.config_files=/Users/sebastiaan/Library/Containers/com.docker.docker/Data/plugins/crazymax_ddext-docs/vm/docker-compose.yml,com.docker.compose.version=2.2.2,com.docker.desktop.plugin=true,com.docker.compose.oneoff=False,org.opencontainers.image.vendor=Docker Inc.,com.docker.compose.image=sha256:c61d6db054191c8b57bd019cc0a4e6b98b8af194226f6b30ff6aae630693553b,com.docker.compose.project.working_dir=/Users/sebastiaan/Library/Containers/com.docker.docker/Data/plugins/crazymax_ddext-docs/vm,com.docker.desktop.plugin.name=DocExt,org.opencontainers.image.title=DocExt","LocalVolumes":"0","Mounts":"","Names":"crazymax_ddext-docs-desktop-extension-service","Networks":"crazymax_ddext-docs-desktop-extension_default","Ports":"80/tcp, 0.0.0.0:8123-\u003e4000/tcp","RunningFor":"2 months ago","Size":"2B (virtual 99.5MB)","State":"running","Status":"Up 4 days"}
```

That's not very "readable", _BUT_ one thing that helps is that it's 1 line per container. This means, I can do things like;

```bash
docker ps --format '{{ json .}}' | grep fedora
{"Command":"\"/bin/bash\"","CreatedAt":"2022-03-08 09:29:44 +0100 CET","ID":"059a153f844c","Image":"fedora:35","Labels":"maintainer=Clement Verna \u003ccverna@fedoraproject.org\u003e","LocalVolumes":"0","Mounts":"","Names":"stupefied_kirch","Networks":"bridge","Ports":"","RunningFor":"2 hours ago","Size":"553MB (virtual 706MB)","State":"running","Status":"Up 2 hours"}
```

Because `grep` is easy to use/remember, whereas it's difficult to remember what options to use for `--filter`, or what options to use for `jq` to "query results".

In this case, I can still "pretty print" the output with `jq` by just piping it;

```bash
docker ps --format '{{ json .}}' | grep fedora | jq
{
  "Command": "\"/bin/bash\"",
  "CreatedAt": "2022-03-08 09:29:44 +0100 CET",
  "ID": "059a153f844c",
  "Image": "fedora:35",
  "Labels": "maintainer=Clement Verna <cverna@fedoraproject.org>",
  "LocalVolumes": "0",
  "Mounts": "",
  "Names": "stupefied_kirch",
  "Networks": "bridge",
  "Ports": "",
  "RunningFor": "2 hours ago",
  "Size": "553MB (virtual 706MB)",
  "State": "running",
  "Status": "Up 2 hours"
}
```

Or extract just a single field with it (`.SomeField` is something that's );

```bash
docker ps --format '{{ json .}}' | grep fedora | jq .CreatedAt
"2022-03-08 09:29:44 +0100 CET"
```

Reply:
Given that JSON is not for humans, I would be tempted to say the default should be the _reverse_ (so never pretty-print by default). Unfortunately, we already have (e.g.) `docker inspect` doing a pretty(ish) print by default, but we could (I think I commented that above) say that _if you (explicitly) pass `--format=json`_ that that always prints "non-pretty".

Reply:
Actually this is already the case:

```console
❯ ./build/docker context ls --format=json
{"Current":true,"Description":"Current DOCKER_HOST based configuration","DockerEndpoint":"unix:///var/run/docker.sock","KubernetesEndpoint":"","Name":"default"}
{"Current":false,"Description":"","DockerEndpoint":"tcp://localhost:12375","KubernetesEndpoint":"","Name":"dind"}
```

```console
❯ ./build/docker context ls --format='{{ json .}}'
{"Current":true,"Description":"Current DOCKER_HOST based configuration","DockerEndpoint":"unix:///var/run/docker.sock","KubernetesEndpoint":"","Name":"default"}
{"Current":false,"Description":"","DockerEndpoint":"tcp://localhost:12375","KubernetesEndpoint":"","Name":"dind"}
```

Reply:
Done, all `inspect` commands will output a compact JSON representation of the elements with `--format=json`